compiler/src/dotty/tools/dotc/core/Definitions.scala

```diff
@@ -1108,6 +1108,22 @@ class Definitions {
     //  - .linkedClass: the ClassSymbol of the enumeration (class E)
     sym.owner.linkedClass.typeRef
 
+  object FunctionTypeOfMethod {
+    /** Matches a `FunctionN[...]`/`ContextFunctionN[...]` or refined `PolyFunction`/`FunctionN[...]`/`ContextFunctionN[...]`.
+     *  Extracts the method type type and apply info.
+     */
+    def unapply(ft: Type)(using Context): Option[MethodOrPoly] = {
+      ft match
+        case RefinedType(parent, nme.apply, mt: MethodOrPoly)
```

Comment:
I'd try to optimize it further by inlining the FunctionOf extractor and specializing it to this use case.
Also, maybe split the `RefinedType` match into two, one for `MethodType` refinements and the other for `PolyType` refinements?